model_analyzer/triton/server.py

```diff
@@ -0,0 +1,309 @@
+            'min-supported-compute-capability' : None,
+            
+            # Backend config
+            'backend-directory' : None,
+            'backend-config' : None,
+            
+            'allow-soft-placement' : None,
+            'gpu-memory-fraction' : None,
+            'tensorflow-version' : None
+        }
+
+    def to_cli_string(self):
+        """
+        Utility function to convert a config into a
+        string of arguments to the server with CLI.
+        """
+        return ' '.join(['--{}={}'.format(key,val) if val else '' \
+                        for key,val in self._server_args.items()])
+
+    def __getitem__(self, key):
+        return self._server_args[key]
+    
+    def __setitem__(self, key, value):
+        if key in self._server_args:
+            self._server_args[key] = value
+        else:
+            raise Exception("The argument '{}' to the Triton Inference Server"
+                             "is not supported by the model analyzer.".format(key))
+    
+class TritonServerFactory:
+    """
+    A factory for creating TritonServer instances
+    """
+
+    def __init__(self):
+        pass
+
+    def create_server(self, run_type, model_path, version, config):
+        if run_type is 'docker':
+            return TritonServerDocker(version=version, model_path=model_path, config=config)
+        elif run_type is 'local':
+            return TritonServerLocal(version=version, model_path=model_path, config=config)
+        else:
+            raise Exception("The run environment '{}' for Triton Inference Server"
+                            "is not supported by the model analyzer.".format(run_type))
+
+class TritonServer:
+    """
+    Defines the interface for the objects created by 
+    TritonServerFactory
+    """
+    def __init__(self, version, model_path, config):
+        self._version = version
+        self._server_config = config
+        self._model_path = model_path
+        assert self._server_config['model-repository'], \
+            "Triton Server requires --model-repository argument to be set."    
+
+        self._logs = None
+
+    def start(self):
+        """
+        Starts the tritonserver 
+        """
+        raise NotImplementedError()
+    
+    def stop(self):
+        """
+        Stops and cleans up after the server
+        """
+        raise NotImplementedError()
+
+    def wait_for_ready(self, num_retries=WAIT_FOR_READY_NUM_RETRIES):
+
+        # FIXME this should not really be a server function
+        if self._server_config['allow-http'] is not False:
+            http_port = self._server_config['http-port'] or 8000
+            url = "http://localhost:{}/v2/health/ready".format(http_port)
+        else:
+            # FIXME to use GRPC to check for ready also
+            raise Exception('allow-http must be True in order to use wait_for_server_ready')
+        
+        retries = num_retries
+        # poll ready endpoint for number of retries
```

Comment:
add empty line before comments, unless comment is the first thing after a structural line (def, if, else, etc.)